#use-added-syntax(tests)
defpackage promises/test:
  import core
  import collections
  import promises

deftest pend-then-resolve:
  val awaitable = 
    within promise = awaitable():
      pend(promise)
      resolve(promise, "fulfilled")
  val promise = make-promise(awaitable)
  #ASSERT(poll(promise) is PromisePending)
  val result = poll(promise) as PromiseResolved
  #ASSERT(value(result) as String == "fulfilled")

deftest thrown-exception-is-rejected:
  val awaitable = 
    within promise = awaitable():
      throw(Exception("Promise rejected"))
  val promise = make-promise(awaitable)
  #ASSERT(poll(promise) is PromiseRejected)

deftest async-await-sanity-check:
  val inner =
    within await = async():
      "hello, async"

  val outer = 
    within await = async():
      #ASSERT(await(inner) as String == "hello, async")
  
  val promise = make-promise(outer)
  #ASSERT(poll(promise) is PromiseResolved)
  